README.md
# send-ranges

[![npm version](http://img.shields.io/npm/v/send-ranges.svg?style=flat-square)](https://www.npmjs.com/package/send-ranges)

Express middleware for handling HTTP range requests. Requires node 18 or higher.

## Installation

```
npm install --save send-ranges
```

## Usage

```js
const path = require('path')
const fse = require('fs-extra')
const express = require('express')
const sendRanges = require('send-ranges')
const app = express()

const mp3Path = path.join(__dirname, 'mp3s')

async function retrieveFile(request) {
  const filename = request.params.filename
  if (!/\.mp3$/.test(filename)) {
    return null // Falsey values will call the next handler in line
  }

  const filePath = path.join(mp3Path, filename)
  const getStream = (range) => fs.createReadStream(filePath, range)
  const type = 'audio/mpeg'
  const stats = await fse.stat(filePath)

  return {getStream, type, size: stats.size}
}

app.get('/:filename', sendRanges(retrieveFile), (req, res) => {
  // If we got here, this was not a range request, or the `retrieveFile` handler
  // returned a falsey value
  res.sendFile(path.join(mp3Path, req.params.filename))
})

app.listen(3000, () => {
  console.log('Example app listening on port 3000!')
})
```

## Options

Exported function accept an `options` config object that will be passed to
`range-parser`. In addition that, this `options` object accept some
`send-ranges` specific ones:

### Pre-send hook

You can pass a `beforeSend` function as an option. This is handy if you want to:

- Handle the response sending yourself
- Use metadata fetched in the file retriever to set headers on the response
- Validate and possibly cancel the response before it is sent

The `retrieveFile` function can return an additional `metadata` property which will be passed to the
`beforeSend` function. Here's an example:

```js
async function retrieveFile(request) {
  const file = await someObjectStore.getFile(request.path)
  const size = file.size
  const metadata = file.metadata
  const type = metadata.contentType
  const getStream = (range) => file.createReadStream(range)
  return {getStream, size, metadata, type}
}

async function beforeSend(info, cb) {
  const {request, response, metadata, sourceStream} = info
  response.set('Last-Modified', new Date(metadata.updated).toUTCString())
  cb()
}

app.get('/*', sendRanges(retrieveFile, {beforeSend}))
```

### Maximum number of ranges

By default, up to two ranges are allowed. The limit can be adjusted with the `maxRanges` option:

```js
sendRanges(retrieveFile, {maxRanges: 4})
```

### intersectRanges

Optional function that allow to modify the list of ranges to be returned. This
can be used for example if the data is not fully available because your file has
holes in it. It will be called with an object with the same `metadata` field
used for the `beforeSend()` optional function, and a `ranges` field with the
list of ranges.

## License

MIT © [Espen Hovlandsdal](https://espen.codes/)


CHANGELOG.md
<!-- markdownlint-disable --><!-- textlint-disable -->

# 📓 Changelog

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [5.0.0](https://github.com/rexxars/send-ranges/compare/v4.0.0...v5.0.0) (2023-09-28)

### ⚠ BREAKING CHANGES

- ship with official typescript types
- require node 18 or higher

### Features

- require node 18 or higher ([1cea745](https://github.com/rexxars/send-ranges/commit/1cea7458a7a11e80fadeecd8b4918ee730df7abc))
- ship with official typescript types ([e7a0d75](https://github.com/rexxars/send-ranges/commit/e7a0d7596e9f9daa303b7ca5efa9cc588d476737))

# Change Log

All notable changes will be documented in this file.

## 4.0.0 - 2020-07-16

### BREAKING

- Drop support for Node 8.

## 3.0.0 - 2019-03-23

### BREAKING

- Drop support for Node 6.

## 2.0.0 - 2018-10-05

### BREAKING

- `Vary: Range` is no longer sent on 206 responses. As long as a cache behaves correctly on a 206 response (according to RFC 2616), this should be a safe (and more correct) behavior.


test/send-ranges.test.js
const fs = require('fs')
const path = require('path')
const pump = require('pump')
const sprom = require('sprom')
const needle = require('needle')
const express = require('express')
const request = require('supertest')
const sendRanges = require('../src')

let server
const fixturePath = path.join(__dirname, 'fixtures', 'text.txt')
const fixtureContent = fs.readFileSync(fixturePath, 'utf8')
const size = fs.statSync(fixturePath).size

const parseResponse = (headers, body) => {
  // Shouldn't make these kind of horrible assumptions on binary/text, headers etc.
  // But this is tests. Controlled environment etc.
  const contentType = headers['content-type']
  const boundary = contentType.slice(contentType.indexOf('boundary=') + 9)

  const content = body.toString()
  const chunks = content
    .split(boundary)
    .map((part) => part.replace(/^-+/, '').trim())
    .map((part) => part.split('\r\n\r\n')[1])
    .filter(Boolean)

  return chunks
}

const getStream = (range) => fs.createReadStream(fixturePath, range)
const getStreamAsync = (range) =>
  new Promise((resolve) => setImmediate(resolve, fs.createReadStream(fixturePath, range)))
const fetchFile = () => ({getStream, size, type: 'text/plain'})
const fetchNothing = () => null
const fetchFileAsync = () => ({getStream: getStreamAsync, size, type: 'text/plain'})
const fetchFileWithoutType = () => ({getStream, size})
const fetchFileWithMetadata = () => ({getStream, size, type: 'text/plain', metadata: {foo: 'bar'}})

const defaultSender = (req, res, next) =>
  pump(getStream({}), res.set({'Content-Length': size, 'Content-Type': 'text/plain'}))

const getApp = (opts = {}) =>
  express().get('/:filename', sendRanges(opts.fetch || fetchFile, opts), defaultSender)

const getListeningApp = (opts) =>
  new Promise((resolve) => {
    server = getApp(opts).listen(0, '127.0.0.1', () => resolve(server.address().port))
  })

const requestRange = async (range, opts = {}) => {
  const port = await getListeningApp(opts)
  const stream = needle.get(`http://127.0.0.1:${port}/somefile.txt`, {
    headers: {Range: range},
  })

  let headers
  stream.on('response', (res) => {
    headers = res.headers
  })

  const body = await sprom.buf(stream)
  const chunks = parseResponse(headers, body)
  const contentLength = parseInt(headers['content-length'], 10)
  return {body, chunks, headers, contentLength}
}

describe('send-ranges', () => {
  afterEach((done) => {
    const close = server ? server.close.bind(server) : setImmediate
    server = null
    close(done)
  })

  test('throws on missing/invalid `getStream`', () => {
    expect(() => sendRanges()).toThrowErrorMatchingSnapshot()
    expect(() => sendRanges('moop')).toThrowErrorMatchingSnapshot()
  })

  test('throws on missing/invalid `beforeSend`', () => {
    expect(() => sendRanges(fetchFile, {beforeSend: null})).toThrowErrorMatchingSnapshot()
    expect(() => sendRanges(fetchFile, {beforeSend: 'moop'})).toThrowErrorMatchingSnapshot()
  })

  test('defers to next handler in chain on no range header', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .expect('Content-Type', /text\/plain/)
      .expect('Content-Length', `${size}`)
      .expect(200, fixtureContent)
      .end(done)
  })

  test('defers to next handler if fetcher returns falsey value', (done) => {
    request(getApp({fetch: fetchNothing}))
      .get('/somefile.txt')
      .set('Range', 'bytes=0-60')
      .expect('Content-Type', /text\/plain/)
      .expect('Content-Length', `${size}`)
      .expect(200, fixtureContent)
      .end(done)
  })

  test('sends 400 on malformed range header', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'foo')
      .expect('Content-Type', /text\/plain/)
      .expect(400, 'Malformed `range` header')
      .end(done)
  })

  test('sends 400 on unsatisfiable range header (unit)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'pages=0-100')
      .expect('Content-Type', /text\/plain/)
      .expect(416, 'Range not satisfiable')
      .end(done)
  })

  test('sends 400 on unsatisfiable range header (invalid numbers)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=moo-mooo')
      .expect('Content-Type', /text\/plain/)
      .expect(416, 'Range not satisfiable')
      .end(done)
  })

  test('sends 400 on unsatisfiable range header (start > end)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=3000-1000')
      .expect('Content-Type', /text\/plain/)
      .expect(416, 'Range not satisfiable')
      .end(done)
  })

  test('sends 400 on unsatisfiable range header (start < 0)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=a0-500000000')
      .expect('Content-Type', /text\/plain/)
      .expect(416, 'Range not satisfiable')
      .end(done)
  })

  test('sends 400 on too many ranges specified', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=0-50,100-150,200-250')
      .expect('Content-Type', /text\/plain/)
      .expect(400, 'Too many ranges specified. Max: 2. Range: bytes=0-50,100-150,200-250')
      .end(done)
  })

  test('sends 400 on too many ranges specified (configured)', (done) => {
    request(getApp({maxRanges: 3}))
      .get('/somefile.txt')
      .set('Range', 'bytes=0-50,100-150,200-250,300-350')
      .expect('Content-Type', /text\/plain/)
      .expect(400, 'Too many ranges specified. Max: 3. Range: bytes=0-50,100-150,200-250,300-350')
      .end(done)
  })

  test('sends single-chunk response on single range', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=0-60')
      .expect('Content-Type', /text\/plain/)
      .expect(206, 'The "Range" header field on a GET request modifies the method')
      .end(done)
  })

  test('sends single-chunk response on single range (async)', (done) => {
    request(getApp({fetch: fetchFileAsync}))
      .get('/somefile.txt')
      .set('Range', 'bytes=0-60')
      .expect('Content-Type', /text\/plain/)
      .expect(206, 'The "Range" header field on a GET request modifies the method')
      .end(done)
  })

  test('sends single-chunk response on single range (head)', (done) => {
    request(getApp())
      .head('/somefile.txt')
      .set('Range', 'bytes=0-60')
      .expect('Content-Type', /text\/plain/)
      .expect(206)
      .end(done)
  })

  test('sends single-chunk response on single range (tail)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=-24')
      .expect('Content-Type', /text\/plain/)
      .expect(206, '(Not Modified) response.')
      .end(done)
  })

  test('sends single-chunk response on single range (tail, open)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=2156-')
      .expect('Content-Type', /text\/plain/)
      .expect(206, '(Not Modified) response.')
      .end(done)
  })

  test('sends single-chunk response on single range (tail, open, no dash)', (done) => {
    request(getApp())
      .get('/somefile.txt')
      .set('Range', 'bytes=2156')
      .expect('Content-Type', /text\/plain/)
      .expect(206, '(Not Modified) response.')
      .end(done)
  })

  test('sends multi-chunk response on multi-range (specific ranges)', async () => {
    const {headers, body, chunks, contentLength} = await requestRange('bytes=0-10,15-25')
    expect(body).toHaveLength(contentLength)
    expect(headers['content-type']).toContain('multipart/byteranges; boundary=')
    expect(chunks).toHaveLength(2)
    expect(chunks[0]).toEqual('The "Range"')
    expect(chunks[1]).toEqual('der field o')
  })

  test('sends multi-chunk response on multi-range (first and last)', async () => {
    const {headers, body, chunks, contentLength} = await requestRange('bytes=0-10,-10')
    expect(body).toHaveLength(contentLength)
    expect(headers['content-type']).toContain('multipart/byteranges; boundary=')
    expect(chunks).toHaveLength(2)
    expect(chunks[0]).toEqual('The "Range"')
    expect(chunks[1]).toEqual(' response.')
  })

  test('sends multi-chunk response on multi-range (specific, tail)', async () => {
    const {headers, body, chunks, contentLength} = await requestRange('bytes=0-0,2156-')
    expect(body).toHaveLength(contentLength)
    expect(headers['content-type']).toContain('multipart/byteranges; boundary=')
    expect(chunks).toHaveLength(2)
    expect(chunks[0]).toEqual('T')
    expect(chunks[1]).toEqual('(Not Modified) response.')
  })

  test('sends multi-chunk response on multi-range (specific ranges)', async () => {
    const opts = {fetch: fetchFileAsync}
    const {headers, body, chunks, contentLength} = await requestRange('bytes=0-10,15-25', opts)
    expect(body).toHaveLength(contentLength)
    expect(headers['content-type']).toContain('multipart/byteranges; boundary=')
    expect(chunks).toHaveLength(2)
    expect(chunks[0]).toEqual('The "Range"')
    expect(chunks[1]).toEqual('der field o')
  })

  // todo: weak etag, last-modified
  test('sends 304 on range & valid if-range', (done) => {
    const eTag = '"foo"'
    const fetchFileWithEtag = () => ({...fetchFile(), eTag: eTag})
    const app = getApp({fetch: fetchFileWithEtag})

    request(app)
      .get('/somefile.txt')
      .set('Range', 'bytes=0-50')
      .set('If-Range', eTag)
      .expect('ETag', eTag)
      .expect(304)
      .end(done)
  })

  test('calls error handler on fetch fail', (done) => {
    const error = new Error('Some error')
    const app = getApp({fetch: () => Promise.reject(error)})
    app.use((err, req, res, next) => {
      expect(err).toEqual(error)
      res.status(404).json({error: 'File not found'})
    })

    request(app)
      .get('/somefile.txt')
      .set('Range', 'bytes=0-100')
      .expect('Content-Type', /application\/json/)
      .expect(404, {error: 'File not found'})
      .end(done)
  })

  test('sets type to octet-stream if no type given', () =>
    request(getApp({fetch: fetchFileWithoutType}))
      .get('/somefile.txt')
      .set('Range', 'bytes=5-26')
      .expect('Content-Type', /application\/octet-stream/)
      .expect(206)
      .then((res) => expect(res.body.toString('utf8')).toEqual('Range" header field on')))

  test('can pass `beforeSend` function, receives metadata, can set headers', () => {
    const beforeSend = (info, cb) => {
      info.response.set('X-Foo', 'some-value')
      expect(info.metadata).toMatchObject({foo: 'bar'})
      setImmediate(cb)
    }
    return request(getApp({beforeSend, fetch: fetchFileWithMetadata}))
      .get('/somefile.txt')
      .set('Range', 'bytes=5-26')
      .expect(206, 'Range" header field on')
      .then((res) => expect(res.headers).toHaveProperty('x-foo', 'some-value'))
  })

  test('calls error handler on `beforeSend` fail', (done) => {
    const error = new Error('Some error')
    const beforeSend = (info, cb) => setImmediate(cb, error)
    const app = getApp({beforeSend})
    app.use((err, req, res, next) => {
      expect(err).toEqual(error)
      res.status(500).json({error: 'Some error'})
    })

    request(app)
      .get('/somefile.txt')
      .set('Range', 'bytes=0-100')
      .expect('Content-Type', /application\/json/)
      .expect(500, {error: 'Some error'})
      .end(done)
  })

  test('can pass `intersectRanges` function', () => {
    const intersectRanges = ({metadata, ranges}) => {
      expect(metadata).toMatchObject({foo: 'bar'})
      return ranges
    }
    return request(getApp({intersectRanges, fetch: fetchFileWithMetadata}))
      .get('/somefile.txt')
      .set('Range', 'bytes=5-26')
      .expect(206, 'Range" header field on')
  })
})


src/index.d.ts
/// <reference types="node" />
import type {NextFunction, Request as ExpressRequest, Response as ExpressResponse} from 'express'
import type {Readable} from 'node:stream'

type Range = {
  start: number
  end: number
}

type StreamFetcher = (range: Range) => Readable | Promise<Readable>

type FileResult<M> = {
  getStream: StreamFetcher
  size: number
  metadata: M
  type?: string
}

type FileFetcher = <M = any>(
  req: Request,
) => FileResult<M> | undefined | null | Promise<FileResult<M> | undefined | null>

interface SendRangesOptions<M = any> {
  beforeSend?: (info: {
    request: ExpressRequest
    response: ExpressResponse
    metadata: M
    sourceStream: Readable
  }) => {}
  maxRanges?: number
  intersectRanges?: (info: {Metadata: M; ranges: Range[]}) => Range[]
}

declare function sendRanges<M = any>(
  fetchStream: FileFetcher,
  opts?: SendRangesOptions<M>,
): (req: ExpressRequest, res: ExpressResponse, next: NextFunction) => Promise<void>

export = sendRanges


example/example.js
/* eslint-disable no-console, no-sync */
const fs = require('fs')
const path = require('path')
const express = require('express')
const sendRanges = require('../')
const app = express()

const filePath = path.join(__dirname, 'somefile.txt')

const retrieveFile = (request) => {
  const filename = request.params.filename
  if (filename !== 'somefile.txt') {
    // We are using promises, so throwing errors will trigger the error middleware
    throw new Error('File not found')
  }

  const getStream = (range) => fs.createReadStream(filePath, range)
  const size = fs.statSync(filePath).size
  const type = 'text/plain'

  // Note that promises are also allowed
  return {getStream, size, type}
}

app.get('/:filename', sendRanges(retrieveFile), (req, res) => {
  // If we got here, this was not a range request
  res.sendFile(filePath)
})

app.listen(3000, () => {
  console.log('Example app listening on port 3000!')
})


src/index.js
const pump = require('pump')
const rangeParser = require('range-parser')
const BRS = require('byte-range-stream')
const noopNext = (info, next) => next()

const defaultOptions = {
  beforeSend: noopNext,
  maxRanges: 2,
}

module.exports = (fetchStream, opts = {}) => {
  const options = {...defaultOptions, ...opts}

  if (typeof fetchStream !== 'function') {
    throw new Error('`fetchStream` must be a function')
  }

  if (typeof options.beforeSend !== 'function') {
    throw new Error('`options.beforeSend` must be a function')
  }

  return async (req, res, next) => {
    // Always flag that we accept ranges
    res.setHeader('Accept-Ranges', 'bytes')

    // Not a range request? Let the usual middleware handle it
    const {range} = req.headers
    if (!range) {
      next()
      return
    }

    // Fetch the file (if we can)
    let file
    try {
      file = await fetchStream(req)
    } catch (err) {
      next(err)
      return
    }

    // If nothing is returned, call the next middleware in line
    if (!file) {
      next()
      return
    }

    const {getStream, size, type, metadata} = file

    // Parse the range header
    let ranges = rangeParser(size, range, options)

    // Malformed?
    if (ranges === -2) {
      res.status(400).type('text').send('Malformed `range` header')
      return
    }

    // Unsatisfiable?
    let isUnsatisfiable = ranges === -1 || ranges.type !== 'bytes'

    const {intersectRanges} = options
    if (!isUnsatisfiable && intersectRanges) {
      ranges = intersectRanges({metadata, ranges})

      isUnsatisfiable = ranges.length === 0
    }

    if (isUnsatisfiable) {
      res
        .status(416)
        .set('Content-Range', `bytes */${size}`)
        .type('text')
        .send('Range not satisfiable')
      return
    }

    // Beyond limit?
    if (ranges.length > options.maxRanges) {
      res
        .status(400)
        .set('Content-Range', `bytes */${size}`)
        .type('text')
        .send(`Too many ranges specified. Max: ${options.maxRanges}. Range: ${range}`)
      return
    }

    // Valid ranges, so flag it as a partial response
    res.status(206)

    // For single-range range requests, we don't need to use multipart
    const isSingleRange = ranges.length === 1
    const resolveStream = isSingleRange
      ? getStream(ranges[0])
      : new BRS({range, getChunk: getStream, totalSize: size, contentType: type})

    // Stream retrieval might be async
    const sourceStream = await resolveStream

    // Prepare headers to set (only apply them after `beforeSend`, in case we run into trouble)
    let headers
    if (isSingleRange) {
      headers = {
        'Content-Type': type || 'application/octet-stream',
        'Content-Range': `bytes ${ranges[0].start}-${ranges[0].end}/${size}`,
        'Content-Length': 1 + ranges[0].end - ranges[0].start,
      }
    } else {
      headers = sourceStream.getHeaders()
    }

    // Allow the user to do pre-response actions, like adding additional headers or even
    // handling the response themselves. The user is responsible for calling next() to
    // continue with the response. Passing an error calls the error handling chain.
    const info = {request: req, response: res, metadata, sourceStream}
    options.beforeSend(info, (err) => {
      if (err) {
        next(err)
        return
      }

      res.set(headers)

      // We don't need the actual body in case of a HEAD, so terminate early
      if (req.method === 'HEAD') {
        res.end()
        return
      }

      // Stream the response!
      pump(sourceStream, res)
    })
  }
}
